add first manuals/networking version (#14)
nitindermohan/oak-new-docs: content/docs/manuals/networking-internals/semantic-addressing.md

  noindex: false # false (default) or true
---

The load balancing in Oakestra consists of a multistep address conversion protocol using semantic addresses as part
of an overlay network.

### Semantic Addressing

The semantic addressing happens on the service layer.
Similar to a cluster IP in Kubernetes, these addresses reference all the instances (replicas) of a microservice
with a single address. This address does not change when scaling up the instances or when migrating them.
Unlike the Kubernetes cluster IP, when deploying a service in Oakestra, the platform provides as many Service IP
addresses as the number of balancing policies supported (and active) within the platform.

#### Example

![Network Example - Dual Stack](_net_example_dual_stack.png)

A developer can communicate with any instance of Service B either with **Round Robin** balancing policy or
the **Closest** balancing policy. The former balances the traffic evenly between all the instances, and the
latter finds the geographically closer instance.

In the following tabs we go into further detail of what is happening with each above shown request.
The ordering goes from top (First) to bottom (Fifth).

> {{< tabs "Requests" >}}

Reviewer: What if in addition to FIrst etc we also add the envelope icon?
<img width="668" alt="image" src="https://github.com/user-attachments/assets/c70bd9a9-a862-4c63-a283-b0a9c612dd69">

I just pushed a new shortcode that allows you to provide an icon together with the tab name, it's called tabs-icon.

```suggestion
{{< tabs-icon "Requests" >}}
```

But I'm missing the icon of the last envelope, can you add it together with the .drawio file you used?

Author: I was also thinking of this, but didn't delve into how to realise it, so thanks for the shortcode! Will definitely use this.

> {{< tab "First" >}}

Reviewer: ```suggestion
{{< tab-icon "First" "/network/NetArchExample_evenlope_1.png">}}
```

Service A performs the first request using the ServiceIP `fdff:1000::1` representing the closest instance balancing policy.
The network components' proxy converts the address to the Namespace IP of Service B Instance 1, which looks like it is
the geographically closer service. The message will be, therefore **transparently** delivered to the closest instance
of Service B. Note that the Namespace IP is the real address of the instance, the one provisioned at deployment time,
unique and dynamic. An application developer never sees or uses this address, as it depends on the subnetwork of the
specific machine where the service is deployed.

> {{< /tab >}}

Reviewer: ```suggestion
{{< /tab-icon >}}
```

> {{< tab "Second">}}

Reviewer: ```suggestion
{{< tab-icon "Second" "/network/NetArchExample_evenlope_2.png">}}
```

Service A performs a second request using the ServiceIP fdff:2000::1 representing the Round Robin balancing
policy. The network components' proxy converts the address to the Namespace IP of Service B Instance 3, which is
randomly chosen among all the available instances.

> {{< /tab >}}

Reviewer: ```suggestion
{{< /tab-icon >}}
```

> {{< tab "Third" >}}

Reviewer: ```suggestion
{{< tab-icon "Third" "/network/NetArchExample_evenlope_3.png">}}
```

When Service A performs a third request using again the ServiceIP fdff:2000::1 representing the Round Robin balancing
policy, this time, the network components' proxy randomly chooses Service B Instance 2.

> {{< /tab >}}

Reviewer: ```suggestion
{{< /tab-icon >}}
```

> {{< tab "Fourth" >}}

Reviewer: ```suggestion
{{< tab-icon "Fourth" "/network/NetArchExample_evenlope_4.png">}}
```

The fourth request from Service A to Service B uses the Instance IP representing instance 3.

The proxy component then automatically chooses Service B Instance 3.

> {{< /tab >}}

Reviewer: ```suggestion
{{< /tab-icon >}}
```

> {{< tab "Fifth" >}}

Reviewer: ```suggestion
{{< tab-icon "Fifth" "/network/NetArchExample_evenlope_4.png">}}
```
In this case we need the icon for the fifth envelope you created.

> The last request demonstrates that IPv4 requests work the same way as described before.

Reviewer: ```suggestion
The last request demonstrates how the requests using IPv4 RoundRobin semantic address works exactly the same way as the IPv6 RoundRobin semantic address used in <img src="{{ $tab.Icon }}" alt= "envelope_1" width="30">.
```

> {{< /tab >}}

Reviewer: ```suggestion
{{< /tab-icon >}}
```

> {{< /tabs >}}

Reviewer: ```suggestion
{{< /tabs-icon >}}
```
